utils/cryptography.go

```diff
@@ -0,0 +1,325 @@
+package utils
+
+// TOMORROWS LABOUR CLEAN, ADD, REFACTOR TEST
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/ecdh"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"math/big"
+	"slices"
+)
+
+type KeyUse int
+
+var (
+	ECDSA KeyUse = 1
+	ECDH  KeyUse = 2
+)
+
+type JWK struct {
+	Key_ops []string `json:"use,omitempty"` // ["sign", "verify", "encrypt", "decrypt", "wrapKey", "unwrapKey", "deriveKey", "deriveBits"]
+	Kty     string   `json:"kty,omitempty"` // "EC", "RSA"
+	Kid     string   `json:"kid,omitempty"` // Key ID
+	Crv     string   `json:"crv,omitempty"` // "P-256"
+	X       string   `json:"x,omitempty"`   // x coordinate as base64 URL encoded string.
+	Y       string   `json:"y,omitempty"`   // y coordinate as base64 URL encoded string.
+	D       string   `json:"d,omitempty"`   // d coordinate as base64 URL encoded string. Private keys only.
+}
+
+// This will give you a private and public key pair both as *JWK structs.
+// The Kid (i.e., "key-id") is equivalent shared but for the prefix "priv_" or "pub_"
+// appended to a random base64 URL encoded string.The Crv parameter is hardcoded
+// for now during key creation as are the key_ops.
+func GenerateKeyPair(keyUse KeyUse) (*JWK, *JWK, error) {
+	id := make([]byte, 16)
+	rand.Read(id)
+	id_str := base64.URLEncoding.EncodeToString(id)
+
+	var privKey JWK
+	privKey.Kty = "EC"
+	privKey.Crv = "P-256"
+	privKey.Kid = "priv_" + id_str
+	privKey.Key_ops = []string{}
+	if keyUse == ECDSA {
+		privKey.Key_ops = append(privKey.Key_ops, "sign")
+	} else if keyUse == ECDH {
+		privKey.Key_ops = append(privKey.Key_ops, "deriveKey")
+	} else {
+		return nil, nil, fmt.Errorf("Unrecognized keyUse. Must be 'ECDSA', or 'ECDH.'")
+	}
+
+	privKey_ecdsaPtr, _ := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	privKey.D = base64.URLEncoding.EncodeToString(privKey_ecdsaPtr.D.Bytes())
+	privKey.X = base64.URLEncoding.EncodeToString(privKey_ecdsaPtr.X.Bytes())
+	privKey.Y = base64.URLEncoding.EncodeToString(privKey_ecdsaPtr.Y.Bytes())
+
+	var pubKey JWK
+	pubKey.Kid = "pub_" + id_str
+	pubKey.Kty = "EC"
+	pubKey.Crv = "P-256"
+	pubKey.Kid = "pub_" + id_str
+	pubKey.Key_ops = []string{}
+	if keyUse == ECDSA {
+		pubKey.Key_ops = append(pubKey.Key_ops, "verify")
+	} else if keyUse == ECDH {
+		pubKey.Key_ops = append(pubKey.Key_ops, "deriveKey")
+	} else {
+		return nil, nil, fmt.Errorf("Unrecognized keyUse. Must be 'ECDSA', or 'ECDH.'")
+	}
+
+	pubKey.X = base64.URLEncoding.EncodeToString(privKey_ecdsaPtr.X.Bytes())
+	pubKey.Y = base64.URLEncoding.EncodeToString(privKey_ecdsaPtr.Y.Bytes())
+
+	return &privKey, &pubKey, nil
+}
+
+func (jwk *JWK) ExportKeyAsGoType() (interface{}, error) {
```

Comment:
If we implement `pri, pub, _ := jwk.ExportKeyAsGoType()`, we will still need an explicit type check due to *ecdsa.key != *ecdh.key. 

Also, after coming over the wire, I imagine the following being an extremely common use case. 

```
var userPublicKey *JWK 
json.unmarshal( userDataFromHttpHeader, userPublickey)
pubKeyAsGoType, _ := userPublicKey.ExportKeyAsGoType()
```

If we implement where it exports double, it would always look like this: 
```
var userPublicKey *JWK 
json.unmarshal( userDataFromHttpHeader, userPublickey)
_,  pubKeyAsGoType, _ := userPublicKey.ExportKeyAsGoType()
```

Reply:
I'll suggest a utility functions as follows: 
`func (*JWK) ExportECDSAKeyPair () (*ecdsa.PrivateKey, *ecdsa.PublicKey, error)`
and
`func (*JWK) ExportECDHKeyPair () (*ecdh.PrivateKey, *ecdh.PublicKey, error)`

This leaves it up to the developer to invoke the appropriate call.

Reply:
Wait, I just realized that it is not necessary to set `jwk.D = ""`
The code would looks as follows: 

```
privJWK_ecdsa1, pubJWK_ecdsa1, _ := utils.GenerateKeyPair(utils.ECDSA)
data := []byte("SIGN ME!")
signature, _ := privJWK_ecdsa1.SignWithKey(data)
verified, err := pubJWK_ecdsa1.CheckAgainstASN1Signature(signature, data)
```

Type checks and conversion occur _within_ the utility functions. In the case of  *JWK.SignWithKey ( data ), it occurs as follows: 
```
	ecdsaPrivateKey, err := privateKey.ExportKeyAsGoType()
	if err != nil {
		return nil, fmt.Errorf("Unable to call ExportKeyAsGoType on function receiver. Error: %w", err)
	}
	var ecdsaKeyAsGoType *ecdsa.PrivateKey
	if privKey, ok := ecdsaPrivateKey.(*ecdsa.PrivateKey); ok {
		ecdsaKeyAsGoType = privKey
	} else {
		return nil, fmt.Errorf("Receiver, privateKey, of type %T could not be cast as compatible Go type, *ecdsa.PrivateKey.", privateKey)
	}
```